tracker/internal/services/service_search.go

```diff
@@ -239,7 +239,71 @@ func (s *searchService) BreadthFirstSearch(server tracker.SearchService_BreadthF
 }
 
 func (s *searchService) DepthFirstSearch(server tracker.SearchService_DepthFirstSearchServer) error {
-	return api.ErrUnimplemented
+	ctx, cancel := context.WithCancel(server.Context())
+	defer cancel()
+
+	done := ctx.Done()
+	stream := consumeStream(ctx, server)
+
+	root := <-stream
+	stack := []*tracker.SearchRequest{
+		root,
+	}
+	seen := map[string]bool{
+		keyFor(root): true,
+	}
+
+	for length := len(stack); length > 0; length = len(stack) {
+		next := make([]*tracker.SearchRequest, 0)
+
+		// Pop
+		node := stack[length-1]
+		stack = stack[0:length]
+
+		// Explore Node
+		var response *tracker.SearchResponse
+		response, err := s.processRequest(ctx, node)
+		if err != nil {
+			return err
+		}
+
+		// Get adjacent nodes
+		nextBatch, err := transformResponse(response)
+		if err != nil {
+			return err
+		}
+
+		for _, item := range nextBatch {
+			key := keyFor(item)
+			if _, ok := seen[key]; !ok {
+				seen[key] = true
+				next = append(next, item)
+			}
+		}
+
+		// send response to client
+		if err := server.Send(response); err != nil {
+			return err
+		}
+
+		select {
+		case <-done:
+			return nil
+
+		// TODO: What this case is for?
+		// Is it to handle some unexpected request from cmd line when the original request is being processed?
```

Comment:
So, search algorithms can be used in one of two ways.

The first is to traverse a subgraph using a `start` node. We exhaust all paths in the substructure using the specified algorithm. Clients are able to interpret the response based on their input (dfs, infix ordering).

The second is to search for an `end` (or set of `ends`) given a `start`. In this case, you want to be able to short-circuit your traversal when all ends have been found.

This cancel block allows clients to gracefully stop a traversal in the case of the latter.

Reply:
yeah I get the use-cases. But I thought, handling cancellation signals on the `done` channel (in first `case`) is meant to cover short-circuiting from the client. Isn't it so?

Reply:
> From a code review standpoint, this looks good. I've been thinking about how to unit test gRPC stream APIs. I'm not a fan of the "proxy" pattern I put into the gateway either and have been exploring other options there.

Tested it now and results look good to me. I think, PR is good to go.

Reply:
Okay I think, I get it now. I miss the point that, we haven't implemented a search client that accepts an end node, which is where the `cancel` flag will be useful